Firestore/core/src/firebase/firestore/util/error_apple.h

```diff
@@ -38,48 +43,45 @@ namespace firestore {
 namespace util {
 
 // Translates a set of error_code and error_msg to an NSError.
-inline NSError* MakeNSError(const int64_t error_code,
-                            const absl::string_view error_msg,
-                            NSError* cause = nil) {
-  if (error_code == FirestoreErrorCode::Ok) {
-    return nil;
-  }
-
-  NSMutableDictionary<NSString*, id>* user_info =
-      [NSMutableDictionary dictionary];
-  user_info[NSLocalizedDescriptionKey] = WrapNSString(error_msg);
-  if (cause) {
-    user_info[NSUnderlyingErrorKey] = cause;
-  }
-
-  return [NSError errorWithDomain:FIRFirestoreErrorDomain
-                             code:static_cast<NSInteger>(error_code)
-                         userInfo:user_info];
-}
+NSError* MakeNSError(const int64_t error_code,
+                     const absl::string_view error_msg,
+                     NSError* cause = nil);
 
 inline NSError* MakeNSError(const util::Status& status) {
   return status.ToNSError();
 }
 
-inline Status MakeStatus(NSError* error) {
-  if (!error) {
-    return Status::OK();
-  }
+Status MakeStatus(NSError* error);
+
+using VoidErrorBlock = void (^)(NSError* _Nullable error);
+
+util::StatusCallback MakeCallback(VoidErrorBlock _Nullable block);
 
-  HARD_ASSERT(error.domain == FIRFirestoreErrorDomain,
-              "Can only translate a Firestore error to a status");
-  auto error_code = static_cast<int>(error.code);
-  HARD_ASSERT(error_code >= FirestoreErrorCode::Cancelled &&
-                  error_code <= FirestoreErrorCode::Unauthenticated,
-              "Unknown error code");
-  return Status{static_cast<FirestoreErrorCode>(error_code),
-                MakeString(error.localizedDescription)};
+template <typename T>
+using VoidValueErrorBlock = void (^)(T _Nullable value,
+                                     NSError* _Nullable error);
+
+template <typename T>
+util::StatusOrCallback<T> MakeCallback(VoidValueErrorBlock<T> _Nullable block) {
+  if (block) {
+    return [block](StatusOr<T> maybe_value) {
+      if (maybe_value.ok()) {
+        block(std::move(maybe_value).ValueOrDie(), nil);
+      } else {
+        block(nil, MakeNSError(std::move(maybe_value).status()));
```

Comment:
FWIW, using a value parameter seems fine to me (though I don't feel very strongly about it):
* this allows transferring ownership of the result back to the caller, which is probably the more common use case;
* the value is independent of its origin, so it saves any thinking about potential lifetime issues.

Michael, for 2 and 3:
2. I don't think there's a reliable optimization for passing parameters (unlike return values), though I might simply be unaware.
3. What case do you have in mind? In this function, `move` is never called twice on the same variable, because those moves are in different branches. Did you have something else in mind? Can you give some example code?

Reply:
I don't feel very strongly either. My thinking is basically that I'd like rules to follow.  And the rule in my head is basically "If a function / method takes a non-primitive value, it should take it as `const &` unless the method does actually need a copy because it wants to locally mutate it or because it wants to hang onto it after the function returns."

There may be exceptions, but I can't clearly articulate why this should be an exceptional case.  In particular I was not sure if this was a performance optimization or a correctness issue or what.  If you can clearly articulate a comment and add it, that would certainly appease me.

And you can ignore my point 3, sorry. You're right it's irrelevant since you can't hit both blocks.

Reply:
Most callbacks are a variation on the make a copy rule. Instead of the callee wanting to make a copy, the caller wants to give away a temporary.

Reply:
> And the rule in my head is basically "If a function / method takes a non-primitive value, it should take it as const & unless the method does actually need a copy because it wants to locally mutate it or because it wants to hang onto it after the function returns."

I'd put it slightly differently, `if a function only needs to "observe" the value, and only for the duration of the function, then use a const reference, otherwise make a copy`. In general, making a copy is safer, so passing by const reference is ultimately an optimization (though an extremely common one and actually useful). In this case, this function invokes a callback, so it doesn't really know what the callback is going to do, and making a copy is the safer choice.

The fact that the callee can optimize the copy by moving the value, in my opinion, doesn't really affect the choice here, except to make passing by value more efficient and consequently more viable.

Reply:
_(I had something written up previously on the topic of passing parameters, mostly concentrating on passing by value. It's not super-tailored to this discussion, but it seems relevant, so I thought I'll just post it here, in case you find something in it useful. Sorry about the wall of text.)_


The idea is:
* if a function only needs to read from a value (by calling const methods on it), it should take the value by const reference to avoid an unnecessary copy;
* if a function needs to take ownership of the value, it can take it by an rvalue reference (`ResourcePath&&`). This avoids an extra copy and forces the caller to be explicit that ownership is passed (they would have to call `std::move` on the parameter, unless the parameter is a temporary);
* (the exception to the previous bullet are move-only types, such as `std::unique_ptr`. In that case, taking it by value or by rvalue reference (`std::unique_ptr<Foo>` vs `std::unique_ptr<Foo>&&`) are identical as far as behavior is concerned, but passing by value is less syntactic noise. I'm saying identical because the lack of a copy constructor will already force the caller to explicitly move the value);

Alternatively, a function taking ownership of the argument can accept the argument by value, which has upsides and downsides. The style guide recommends this approach. The upside is that both lvalues and rvalues can bind to the argument:

1. Consider what happens if the caller wants to still retain the value that they pass as the parameter to our function:
```cpp
void by_value(std::string s);
void by_rvalue_ref(std::string&& s);

std::string s = create_s(); // The string is used after calling by_value/by_rvalue_ref

by_value(s); // by_value gets a copy of s
std::cout << s; // OK -- our copy of s is still valid

by_rvalue_ref(s); // Doesn't compile -- explicit move is required
by_rvalue_ref(std::move(s)); // Now our copy of s is moved from, reading from it would be undefined behavior
std::cout << s; // Bug -- trying to access a moved-from value
// Instead, we can do:
std::string s_copy = s;
by_value_ref(std::move(s_copy));
std::cout << s; // Our copy of s is still valid, but this is ugly
```

2. Another upside is that we only need one function to deal with both lvalues and rvalues:
```cpp
void by_value(std::string s);

std::string s1 = create_s();
std::string s2 = create_s();
const std::string& s3 = get_s();

by_value(s1); // Fine -- a copy is performed, but that's what the caller (presumably) wanted
by_value(std::move(s2)); // Fine
by_value(s3); // Fine -- a caller cannot avoid a copy in this case
by_value(create_s()); // Fine
```

This is especially important when a function has several parameters, each of which should be able to bind to either an lvalue or an rvalue. Writing all possible combinations leads to a combinatorial explosion. Even with just two parameters:
```cpp
void foo(const std::string& s1, const std::string& s2);
void foo(std::string&& s1, const std::string& s2);
void foo(const std::string& s1, std::string&& s2);
void foo(std::string&& s1, std::string&& s2);
// vs
void foo(std::string s1, std::string s2);
```

The downsides of this approach are:
1. The caller has to cooperate:
```cpp
using LargeT = std::vector<std::string>;
void by_value(LargeT v);
void by_rvalue_ref(LargeT&& v);

LargeT v = create_gigabyte_of_text();
// If a caller forgets to call move, the code will compile just fine,
// resulting in a silent unnecessary copy
by_value(v);
by_rvalue_ref(v); // Won't compile, reminding the caller they have to move
```
2. It's slightly less efficient. Whereas `void by_rvalue_ref(std::string&& s)` would perform a single move, `void by_value(std::string s)` would perform two. This is negligible most of the time (at least when dealing with types for which move is more efficient than copying). To illustrate:
```cpp
struct Foo {
  std::string mem_s_;
  void by_value(std::string arg) {
    mem_s_ = std::move(s);
  }
  void by_rvalue_ref(std::string&& arg) {
    mem_s_ = std::move(s);
  }
};

std::string s1, s2;
Foo foo;

foo.by_value(std::move(s1));
foo.by_rvalue_ref(std::move(s2));
```
When `foo.by_value` is called, there are two move operations performed: one move constructor to create `arg` from `s`, and one move assignment operator to assign `arg` to `mem_s_`. However, when `foo.by_rvalue_ref` is called, only the move assignment operator (`arg` to `mem_s_`) is called. Move constructor is not invoked, because in this case `arg` is just a reference (remember, `std::move` itself is just a cast, it doesn't move anything).

Reply:
Thanks for all the explanation.  I agree that passing by const reference is an optimization and it's certainly not important what we do here. Pass by value is fine. Please submit the PR, and feel free to punt on this conversation. It's only an academic exercise at this point.

That said, I still don't understand why this case is different than other cases where we pass by const reference.  

For example, is this code okay?
```
void a(const Foo &foo) {
  b(foo);
}
```

I would say that a() "only needs to 'observe' the value, and only for the duration of the function"...  I don't know what b() needs to do, but that's up to its signature to determine.

I'd say the same argument holds for this code:
```
void forwardStatus(const StatusOr<T> &maybe_value) {
  if (maybe_value.ok()) {
    block(maybe_value.ValueOrDie(), nil);
  } else {
    block(nil, MakeNSError(maybe_value.status()));
  }
}
```

Are we on the same page here?  If so, then I guess the differentiating factor (reason why Gil's code shouldn't pass by `const&`) is that it's a lambda expression?  Is there a "rule" that lambda expressions should always pass-by-value or something?

Reply:
Without knowing the signature of `b` in this example we can't tell if this is the best signature:

```
void a(const Foo &foo) {
  b(foo);
}
```

If `b` takes `foo` by value, then `a` would have been better off also declaring `foo` by value. In this regard your implementation leaks out into your interface.

However, in the case of a callback we have the reverse relationship: the interface is specified for some unknown callers to implement. In this case it's more interesting what the caller of the callback wants to do: is the caller giving out references to retained data? In that case it should be a const reference. Is the caller generating a temporary? Then it should be a value.

In a way the argument to a callback is more analogous to a return value than an argument. When writing the type of a return value you give out const references to data you're retaining, but return by value if you're generating a temporary. There are some practicalities for return values that make this a must-follow rule rather than anything up for debate, but the analogy for why you make the decisions the way you do holds (I think).